prov/verbs/src/windows/verbs_nd_ibv.c

```diff
@@ -0,0 +1,857 @@
+	device_attr->max_qp = VRB_ND_MAX_QP_CNT;
+	device_attr->max_qp_wr = nd_attr->MaxInitiatorQueueDepth;
+	device_attr->max_sge = nd_attr->MaxInitiatorSge;
+	device_attr->max_sge_rd = nd_attr->MaxReadSge;
+	device_attr->max_cq = VRB_ND_MAX_CQ_CNT;
+	device_attr->max_cqe = nd_attr->MaxCompletionQueueDepth;
+	device_attr->max_mr = VRB_ND_MAX_MR_CNT;
+	device_attr->max_qp_rd_atom = nd_attr->MaxInboundReadLimit;
+	device_attr->max_qp_init_rd_atom = nd_attr->MaxOutboundReadLimit;
+	device_attr->max_srq = VRB_ND_MAX_SRQ_CNT;
+	device_attr->max_srq_wr = nd_attr->MaxSharedReceiveQueueDepth;
+	device_attr->phys_port_cnt = 1;
+
+	return FI_SUCCESS;
+}
+
+const char *ibv_get_device_name(struct ibv_device *device)
+{
+	trace(__func__);
+	if (!device)
+		return NULL;
+
+	return device->name;
+}
+
+// infiniband/verbs.h defines ibv_query_port to be ___ibv_query_port
+int ___ibv_query_port(struct ibv_context *context, uint8_t port_num,
+		      struct ibv_port_attr *port_attr)
+{
+	trace(__func__);
+	if ((!context) || (!port_attr)) {
+		errno = EINVAL;
+		return errno;
+	}
+
+	struct adapter_t *adapter = get_adapter_by_context(context);
+	if (!adapter) {
+		errno = ENODEV;
+		return errno;
+	}
+
+	memset(port_attr, 0, sizeof(struct ibv_port_attr));
+	port_attr->state = IBV_PORT_ACTIVE;
+	port_attr->max_mtu = IBV_MTU_4096;
+	port_attr->active_mtu = IBV_MTU_4096;
+	port_attr->gid_tbl_len = 0;
+	port_attr->max_msg_sz = adapter->info.MaxTransferLength;
+	port_attr->pkey_tbl_len = 0;
+	port_attr->lid = 0;
+	port_attr->sm_sl = 0;
+	port_attr->active_width = 2; // 4x
+	port_attr->active_speed = 32; // 25 Gbps
+	port_attr->phys_state =
+		5; // IB_PORT_PHYS_STATE_LINKUP from "infiniband/iba/ib_types.h"
+	port_attr->link_layer = IBV_LINK_LAYER_ETHERNET;
+
+	return FI_SUCCESS;
+}
+
+int ibv_query_gid(struct ibv_context *context, uint8_t port_num, int index,
+		  union ibv_gid *gid)
+{
+	trace(__func__);
+	gid->global.subnet_prefix =
+		0x80fe; // little endian representation of link-local address prefix.
+	gid->global.interface_id = 0;
+	return FI_SUCCESS;
+}
+
+int ibv_query_pkey(struct ibv_context *context, uint8_t port_num, int index,
+		   __be16 *pkey)
+{
+	trace(__func__);
+	*pkey = 0xffff;
+	return FI_SUCCESS;
+}
+
+struct ibv_pd *ibv_alloc_pd(struct ibv_context *context)
+{
+	trace(__func__);
+	if (!context) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	struct ibv_pd *pd = (struct ibv_pd *)calloc(1, sizeof(struct ibv_pd));
+	if (!pd) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	pd->context = context;
+
+	return pd;
+}
+
+int ibv_dealloc_pd(struct ibv_pd *pd)
+{
+	trace(__func__);
+	if (!pd) {
+		errno = EINVAL;
+		return errno;
+	}
+	free(pd);
+	return FI_SUCCESS;
+}
+
+struct ibv_comp_channel *ibv_create_comp_channel(struct ibv_context *context)
+{
+	trace(__func__);
+	if (!context) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	struct ibv_comp_channel *channel = (struct ibv_comp_channel *)calloc(
+		1, sizeof(struct ibv_comp_channel_vrb_nd));
```

Comment:
I'll switch around the places where I allocate memory. That's easy enough. Do you have a preference for how to obtain a pointer to the outer structure when a pointer to the inner structure is passed in as a parameter?

This is very much like a derived class in C++ where a pointer to base is returned. Since there is no dynamic_cast in C to obtain the pointer to derived again, it has to be just a cast. It seemed like I saw container_of being used frequently in this way, but I can just use straight casts if you think that would be clearer.

Reply:
Using container_of to get an outer structure from an inner one is what we want.  The problem here is that the code is allocating the size of the *outer* structure, but assigning the memory to a pointer that's an *inner* structure.  It then uses container_of to get the outer structure.  This flow only works if the inner structure is first member of the outer.  I'm only referring to this flow; there's no actual issue with using container_of.  That's ideal.